## Workflow for further examination of hgcA+ PVC bins


#### Collect PVC references

Scripts for this here: `code/binning/PVC_details/collect_PVC_references.sh`.
For all of these, I gathered the genome fna files (here: `~/HellsCanyon/dataEdited/binAnalysis/PVC_details/genome_dereplication/genomes`).
After gathering them all I predicted the ORFs.

**Dereplicate bins from this project**

I wanted to include both hgcA+ and hgcA- bins from HCC.
I needed to dereplicate these, sine the hgcA+ bins were manually curated, but the rest were autogenerated (which includes some hgcA+ ones).
I used Sarah's ANI workflow for this.
R scripts to clean up the data are here: `PVC_hgcA_minus_derep.R`.


**Collect 5M PVC hgcA+ bins**

All hgcA+ bins from the 5M project on Lake Mendota were included.

**Bins from Jones et al, 2019**

PVC bins from Dan Jones's 2020 paper.
I made a preliminary tree, and these were not closely related.
I kept them in, but they aren't very close.

**References selected for 5M project**

Known isolate genomes for both these phyla.
I picked out some to use with the 5M project, which were stored on the GLBRC servers here: `~/references/genomes`.
I copied over the Kiritimatiellaeota and Lentisphaerae ORFs.
Also include two Planctomycetes genomes, for an outgroup (GCF_000255705.1 and GCF_000025185.1).
In our initial tree, there weren't many good RefSeq or even GenBank genomes that were close to our bins.
I looked at the GTDB tree (`PVC_tree.R`)to confirm that there wasn't more genomes to be pulled here, but there weren't many, so I left the reference set as is.

**Additional genomes**

There were a few that I wanted to include that we didn't have previously.
My notes on collecting them are here.
All of them will be added to the reference folder: `~/references/genomes/bins/`

1. *Victivallis vadensis* genome in here, which we know is GCA_003096415.1. Let's download this from NCBI as well, here: https://www.ncbi.nlm.nih.gov/assembly/GCF_003096415.1. Saved it by its accession ID: `GCF_003096415.1_ASM309641v1_genomic.fna`.

2. I also wanted to include the two genomes from van Vliet et al, 2019, in which they isolated two Kiritimatiellaeota organisms. These are listed at ENA under these accession numbers: SAMEA5207384 and SAMEA5207385. I downloaded the genome fasta files from NCBI here:
  - Pontiella desulfatans: https://www.ncbi.nlm.nih.gov/assembly/GCF_900890425.1.
  - Pontiella sulfatireligans: https://www.ncbi.nlm.nih.gov/assembly/GCF_900890705.1

Moved all of these over, cleaned them up.


**Remove scaffolds from HC18HY300_bin_0036**

The tree analysis downstream identified two scaffolds that were present in two genomes (this is why I need to clean up my bin-generating pipelines).
The two bins are anvio_hgcA_0110 and HC18HY300_bin_0036.
We'll remove the two scaffolds (HC18HY300_000000021381 and HC18HY300_000000059556) from HC18HY300_bin_0036.
Now will continue as before.


**Predict ORFs**

Then I ran the IMMA_ORF_STAN workflow to get the needed ORFs.
After it was done, I concatenated the ORFs and generated the gene-to-bin file.



#### hgcA analysis for PVC

**Identify hgcA sequences**

First I used the `protein_identification_and_alignment.py` script in HomeBio to identify potential *hgcA* sequences and saved out an alignment file.
I downloaded this alignment file to my local computer here: `dataEdited/bins/binAnalysis/PVC_details/hgcA_analysis/hgcA_hits/PVC_hgcA.afa`.
I then used the alignment viewer within DECIPHER to look at the alignment with `BrowseSeqs`.
R script is here: `code/binning/PVC_details/hgcA_alignment.R`
All these sequences look like true *hgcA* sequences.
Next I made a list and downloaded it to my local computer (`PVC_hgcA_geneID_list.txt`).
I also pulled out the relevant entries from the G2B file for these *hgcA* sequences (`PVC_hgcA_G2B.tsv`).


**Generate hgcA tree**

I then took the HgcA alignment I generated earlier and masked it at 50% gaps using Trimal.
Then, I used RAxML to generate a maximum likelihood phylogeny.
I downloaded the bipartitions (`RAxML_bipartitions.PVC_hgcA`) and the info (`RAxML_info.PVC_hgcA`) files to my local computer (`dataEdited/bins/binAnalysis/PVC_details/hgcA_analysis/tree_generation`).
This data was aggregated into a tree here: `code/binning/PVC_details/hgcA_tree.R`.
Note, the final version of this includes some bin information that was gathered in the "Tree generation for PVC" step, described below.



#### Tree generation for PVC

Scripts for this are here: `code/binning/PVC_details/tree_generation_for_PVC.sh`


**Run tree generation workflow**

I then ran my tree generation workflow from HomeBio.
I used the rp16 bacteria HMMs requiring hits agains at least 8 of the HMMs.
I masked the alignment at 50% gaps, then ran RAxML on the alignment.
I downloaded these files here: `dataEdited/bins/binAnalysis/PVC_details/tree_generation`.


**Set up dataframe for renaming tips**

I manually made a tsv file with a name for the bin, the accession number, the phylum, and the name of the tip label: `tip_naming.xlsx`.
This is also in the `tree_generation` folder.
I was mostly interested in getting names for the reference sequences here.
The RefSeq genomes (GCF prefix) will be named with their genus and species (when available).
The GenBank genomes (GCA prefix) will be named with their order names, as determined by GTDB.
These two groups will get an accession ID after their name in the tree.



`code/phylogenies/PVC_tree.R`.

First I read out an unrooted tree, just to check it.
I did a few iterations of it this way.
Settled on using some Planctomycetes as outgroup to root this.

I then renamed all the branches using this file: `dataEdited/binning/phylogeny/PVC/tip_naming.tsv`.
I also generated a color vector to color code the bins by source.

I also labeled the Lentisphaerae and Kiritimatiellaeota branches.
I always have a hard time making these actually pretty, so I saved it out as "unedited", and manually cleaned it up in Illustrator.
