## Workflow for further examination of hgcA+ PVC bins


#### Collect PVC references

Scripts for this here: `code/binning/PVC_details/collect_PVC_references.sh`.
For all of these, I gathered the genome fna files (here: `~/HellsCanyon/dataEdited/binAnalysis/PVC_details/genome_dereplication/genomes`).
After gathering them all I predicted the ORFs.

**Dereplicate bins from this project**

I wanted to include both hgcA+ and hgcA- bins from HCC.
I needed to dereplicate these, sine the hgcA+ bins were manually curated, but the rest were autogenerated (which includes some hgcA+ ones).
I used Sarah's ANI workflow for this.
R scripts to clean up the data are here: `PVC_hgcA_minus_derep.R`.


**Collect 5M PVC hgcA+ bins**

All hgcA+ bins from the 5M project on Lake Mendota were included.

**Bins from Jones et al, 2019**

PVC bins from Dan Jones's 2020 paper.
I made a preliminary tree, and these were not closely related.
I kept them in, but they aren't very close.

**References selected for 5M project**

Known isolate genomes for both these phyla.
I picked out some to use with the 5M project, which were stored on the GLBRC servers here: `~/references/genomes`.
I copied over the Kiritimatiellaeota and Lentisphaerae ORFs.
Also include two Planctomycetes genomes, for an outgroup (GCF_000255705.1 and GCF_000025185.1).
In our initial tree, there weren't many good RefSeq or even GenBank genomes that were close to our bins.
I looked at the GTDB tree (`PVC_tree.R`)to confirm that there wasn't more genomes to be pulled here, but there weren't many, so I left the reference set as is.

**Additional genomes**

There were a few that I wanted to include that we didn't have previously.
My notes on collecting them are here.
All of them will be added to the reference folder: `~/references/genomes/bins/`

1. *Victivallis vadensis* genome in here, which we know is GCA_003096415.1. Let's download this from NCBI as well, here: https://www.ncbi.nlm.nih.gov/assembly/GCF_003096415.1. Saved it by its accession ID: `GCF_003096415.1_ASM309641v1_genomic.fna`.

2. I also wanted to include the two genomes from van Vliet et al, 2019, in which they isolated two Kiritimatiellaeota organisms. These are listed at ENA under these accession numbers: SAMEA5207384 and SAMEA5207385. I downloaded the genome fasta files from NCBI here:
  - Pontiella desulfatans: https://www.ncbi.nlm.nih.gov/assembly/GCF_900890425.1.
  - Pontiella sulfatireligans: https://www.ncbi.nlm.nih.gov/assembly/GCF_900890705.1

Moved all of these over, cleaned them up.


**Remove scaffolds from HC18HY300_bin_0036**

The tree analysis downstream identified two scaffolds that were present in two genomes (this is why I need to clean up my bin-generating pipelines).
The two bins are anvio_hgcA_0110 and HC18HY300_bin_0036.
We'll remove the two scaffolds (HC18HY300_000000021381 and HC18HY300_000000059556) from HC18HY300_bin_0036.
Now will continue as before.



**Predict ORFs**

Then I ran the IMMA_ORF_STAN workflow to get the needed ORFs.
After it was done, I concatenated the ORFs and generated the gene-to-bin file.



#### Tree generation for PVC


**Search for rp16 genes**


*Concatenate all bins*

First I concatenated all the bins into a single ORF faa file.
Then I generated a gene to bin file that linked each gene ID to its respective bin.
I searched the concatenated ORF file for each of the rp16 genes using HMMs with hmmsearch (v3.3.1).
I then pulled out the amino acid sequences for the rp16 genes and aligned them.

*Rename fasta headers with bin name*

I then renamed the fasta file headers with the name of the bin so that they could be concatenated later in Geneious.

*Generate alignment*

I then downloaded all the clean alignments to my local computer and imported them into Geneious.
The rpS3 gene has duplicate hits in LEN_0037, so I deleted LEN_0037 2.
I then concatenated all the alignments.
KIR_0040 had less than 1000 ungapped residues and was clearly missing a bunch of the proteins, so I initially removed it from the analysis, but in later iterations kept it, due to its similarity to a few of the genomes from HCC.
I exported this as `rp16_alignment_masked.afa` and uploaded it to the GLBRC servers: `~/HellsCanyon/dataEdited/binning/manualBinning/binsGood/phylogeny/PVC/tree_building`.

**Generate tree**

I then generated a maximum likelihood tree using RAxML (v8.2.11).
The R scripts to prepare this tree are here: `code/phylogenies/PVC_tree.R`.

First I read out an unrooted tree, just to check it.
I did a few iterations of it this way.
Settled on using some Planctomycetes as outgroup to root this.

I then renamed all the branches using this file: `dataEdited/binning/phylogeny/PVC/tip_naming.tsv`.
I also generated a color vector to color code the bins by source.

I also labeled the Lentisphaerae and Kiritimatiellaeota branches.
I always have a hard time making these actually pretty, so I saved it out as "unedited", and manually cleaned it up in Illustrator.


*Set up dataframe for renaming tips*

I manually made a tsv file with a name for the bin, the accession number, the phylum, and the name of the tip label: `dataEdited/binning/phylogeny/PVC/tip_naming.tsv`.
